File: backend/app/routers/review.py
from fastapi import APIRouter, Depends, status, Query
from sqlalchemy.orm import Session
from typing import List

from app.core.database import get_db
from app.core.dependencies import get_current_user, require_admin
from app.models.user import User
from app.schemas.review import (
    ReviewCreate, 
    ReviewUpdate,
    ReviewResponse, 
    BookRatingSummary
)
from app.services import review as review_service

router = APIRouter(prefix="/reviews", tags=["Reviews"])


@router.get("/book/{book_id}", response_model=List[ReviewResponse])
def get_book_reviews(
    book_id: str,
    skip: int = Query(0, ge=0),
    limit: int = Query(50, ge=1, le=100),
    db: Session = Depends(get_db)
):
    """Lấy danh sách đánh giá của sách (Public)"""
    return review_service.get_book_reviews(db, book_id, skip, limit)


@router.get("/book/{book_id}/summary", response_model=BookRatingSummary)
def get_rating_summary(
    book_id: str, 
    db: Session = Depends(get_db)
):
    """Lấy thống kê đánh giá của sách (Public)"""
    return review_service.get_rating_summary(db, book_id)


@router.post("/", response_model=ReviewResponse, status_code=status.HTTP_201_CREATED)
def create_review(
    data: ReviewCreate,
    current_user: User = Depends(get_current_user),
    db: Session = Depends(get_db)
):
    """Tạo đánh giá mới (Authenticated)"""
    return review_service.create_review(db, current_user.user_id, data)


@router.put("/{review_id}", response_model=ReviewResponse)
def update_review(
    review_id: int,
    data: ReviewUpdate,
    current_user: User = Depends(get_current_user),
    db: Session = Depends(get_db)
):
    """Cập nhật đánh giá của mình (Authenticated)"""
    return review_service.update_review(db, review_id, current_user.user_id, data)


@router.delete("/{review_id}", status_code=status.HTTP_204_NO_CONTENT)
def delete_review(
    review_id: int,
    current_user: User = Depends(get_current_user),
    db: Session = Depends(get_db)
):
    """Xóa đánh giá của mình (Authenticated)"""
    review_service.delete_review(db, review_id, current_user.user_id, is_admin=False)
    return None


@router.delete("/admin/{review_id}", status_code=status.HTTP_204_NO_CONTENT)
def delete_review_admin(
    review_id: int,
    current_admin: User = Depends(require_admin),
    db: Session = Depends(get_db)
):
    """Admin xóa đánh giá bất kỳ"""
    review_service.delete_review(db, review_id, current_admin.user_id, is_admin=True)
    return None

File: frontend/src/api/review.ts
const API_BASE_URL = process.env.NEXT_PUBLIC_API_URL || 'http://127.0.0.1:8000';

export interface ReviewResponse {
  review_id: number;
  book_id: string;
  user_id: string;
  user_fullname: string | null;
  rating: number;
  comment: string | null;
  created_at: string;
}

export interface BookRatingSummary {
  average_rating: number;
  total_reviews: number;
  rating_distribution: {
    [key: string]: number;
  };
}

export interface ReviewCreate {
  book_id: string;
  rating: number;
  comment?: string;
}

export interface ReviewUpdate {
  rating?: number;
  comment?: string;
}

// Helper để lấy token
function getAuthToken(): string | null {
  if (typeof window === 'undefined') return null;
  return localStorage.getItem('access_token');
}

// Helper để tạo headers với auth
function getAuthHeaders(): HeadersInit {
  const token = getAuthToken();
  return {
    'Content-Type': 'application/json',
    ...(token ? { 'Authorization': `Bearer ${token}` } : {})
  };
}

/**
 * Lấy danh sách đánh giá của sách (Public)
 */
export async function getBookReviews(
  bookId: string,
  skip: number = 0,
  limit: number = 50
): Promise<ReviewResponse[]> {
  const params = new URLSearchParams();
  params.append('skip', skip.toString());
  params.append('limit', limit.toString());

  const response = await fetch(
    `${API_BASE_URL}/api/reviews/book/${bookId}?${params.toString()}`,
    {
      method: 'GET',
      headers: { 'Content-Type': 'application/json' },
    }
  );

  if (!response.ok) {
    const errorData = await response.json().catch(() => ({}));
    throw new Error(errorData.detail || 'Lấy đánh giá thất bại');
  }

  return response.json();
}

/**
 * Lấy thống kê đánh giá của sách (Public)
 */
export async function getBookRatingSummary(bookId: string): Promise<BookRatingSummary> {
  const response = await fetch(
    `${API_BASE_URL}/api/reviews/book/${bookId}/summary`,
    {
      method: 'GET',
      headers: { 'Content-Type': 'application/json' },
    }
  );

  if (!response.ok) {
    const errorData = await response.json().catch(() => ({}));
    throw new Error(errorData.detail || 'Lấy thống kê thất bại');
  }

  return response.json();
}

/**
 * Tạo đánh giá mới (Authenticated)
 */
export async function createReview(data: ReviewCreate): Promise<ReviewResponse> {
  const response = await fetch(`${API_BASE_URL}/api/reviews/`, {
    method: 'POST',
    headers: getAuthHeaders(),
    body: JSON.stringify(data),
  });

  if (!response.ok) {
    const errorData = await response.json().catch(() => ({}));
    throw new Error(errorData.detail || 'Tạo đánh giá thất bại');
  }

  return response.json();
}

/**
 * Cập nhật đánh giá (Authenticated)
 */
export async function updateReview(
  reviewId: number,
  data: ReviewUpdate
): Promise<ReviewResponse> {
  const response = await fetch(`${API_BASE_URL}/api/reviews/${reviewId}`, {
    method: 'PUT',
    headers: getAuthHeaders(),
    body: JSON.stringify(data),
  });

  if (!response.ok) {
    const errorData = await response.json().catch(() => ({}));
    throw new Error(errorData.detail || 'Cập nhật đánh giá thất bại');
  }

  return response.json();
}

/**
 * Xóa đánh giá của mình (Authenticated)
 */
export async function deleteReview(reviewId: number): Promise<void> {
  const response = await fetch(`${API_BASE_URL}/api/reviews/${reviewId}`, {
    method: 'DELETE',
    headers: getAuthHeaders(),
  });

  if (!response.ok) {
    const errorData = await response.json().catch(() => ({}));
    throw new Error(errorData.detail || 'Xóa đánh giá thất bại');
  }
}

/**
 * Admin xóa đánh giá
 */
export async function deleteReviewAdmin(reviewId: number): Promise<void> {
  const response = await fetch(`${API_BASE_URL}/api/reviews/admin/${reviewId}`, {
    method: 'DELETE',
    headers: getAuthHeaders(),
  });

  if (!response.ok) {
    const errorData = await response.json().catch(() => ({}));
    throw new Error(errorData.detail || 'Xóa đánh giá thất bại');
  }
}